sqlgen/integration_test.go

```diff
@@ -174,6 +196,178 @@ func TestBatchFilter(t *testing.T) {
+				{Species: "Bubo virginianus", CommonName: "Great Horned Owl", Genus: "Bubo", Family: "Strigidae"},
+			},
+		},
+		{
+			name:              "main thread blocks with broad query",
+			description:       "Main thread performs a broad SELECT ... FOR UPDATE. The goroutine does a narrow SELECT ... FOR UPDATE and should be blocked by the main thread's query. The main thread INSERTs a new row, and the goroutine should pick it up.",
+			mainThreadFilter:  Filter{"family": "Strigidae"},
+			mainThreadOptions: &SelectOptions{ForUpdate: true},
+			goroutineFilter:   Filter{"family": "Strigidae", "genus": "Megascops"},
+			goroutineOptions:  &SelectOptions{ForUpdate: true, OrderBy: "species"},
+			mainThreadUpdateFunc: func(ctx context.Context, db *DB, waiter *sync.WaitGroup) error {
+				// INSERT A new row that the goroutine should get in its SELECT query
+				if _, err := db.InsertRow(ctx, &Owl{Species: "Megascops asio", CommonName: "Eastern Screech Owl", Genus: "Megascops", Family: "Strigidae"}); err != nil {
+					return err
+				}
+				return nil
+			},
+			goroutineExpectedResult: []*Owl{
+				{Species: "Megascops asio", CommonName: "Eastern Screech Owl", Genus: "Megascops", Family: "Strigidae"},
+				{Species: "Megascops kennicottii", CommonName: "Western Screech-Owl", Genus: "Megascops", Family: "Strigidae"},
+			},
+		},
+		{
+			name:              "non-blocking queries on different parts of the index",
+			description:       "The main thread and the goroutine both perform a SELECT ... FOR UPDATE, but their queries shouldn't block each other because they are selecting different parts of the index",
+			mainThreadFilter:  Filter{"family": "Strigidae", "genus": "Megascops"},
+			mainThreadOptions: &SelectOptions{ForUpdate: true},
+			goroutineFilter:   Filter{"family": "Strigidae", "genus": "Psiloscops"},
+			goroutineOptions:  &SelectOptions{ForUpdate: true},
+			mainThreadUpdateFunc: func(ctx context.Context, db *DB, waiter *sync.WaitGroup) error {
+				// Wait until the goroutine has performed its SELECT
+				waiter.Wait()
+
+				// INSERT a new row
+				if _, err := db.InsertRow(ctx, &Owl{Species: "Megascops asio", CommonName: "Eastern Screech Owl", Genus: "Megascops", Family: "Strigidae"}); err != nil {
+					return err
+				}
+				return nil
+			},
+			goroutineExpectedResult: []*Owl{
+				{Species: "Psiloscops flammeolus", CommonName: "Flammulated Owl", Genus: "Psiloscops", Family: "Strigidae"},
+			},
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			// Create the test database and insert some initial rows into the owls table
+			tdb, db, err := setup()
+			assert.NoError(t, err)
+			defer tdb.Close()
+
+			err = db.InsertRows(context.Background(), initialDbState, 100)
+			assert.NoError(t, err)
+
+			// Begin a transaction and SELECT ... FOR UPDATE some rows
+			txCtx, tx, err := db.WithTx(context.Background())
```

Comment:
nit: since we declare txCtx is instantiated by both the main thread and goroutine, should we consider giving each context a unique name to disambiguate?

Reply:
I think it's fine because they're scoped differently.  But just to be very explicit/clear, I'll give the goroutine's context and transaction different names.